### ISSUE TITLE
Filling doc gaps for RNNs: example of training vanilla RNN, intro to SequenceGenerator/Readout/etc

### ISSUE THREAD

**Issue by mjwillson** [bartvm/blocks#587]
Hi all

Currently working through the nice docs on RNNs at https://github.com/bartvm/blocks/blob/master/docs/rnn.rst -- could I make a plea for just a little bit more info on how to take a SimpleRecurrent and train it with an output layer and a cost function?

There's a bit of a gap between these docs, which take you as far as generating a sequence of hidden states with a SimpleRecurrent but don't get you to the point of adding an output layer and a cost function and training the thing, and the code examples (markov chain / reverse words) which use more sophisticated models with attention mechanisms and are abstracted more than the examples in the docs, where it's a bit hard for a newbie to figure out what's going on and apply it to a simpler model.

Might I humbly suggest extending the rnn docs page with a bit of info about SequenceGenerator/Readout/Emmitter/Feedback, how they all fit together and how you use them to compile a theano function that computes your cost function for the sequence?

Failing that if anyone has any simpler example code they could point me at for training a more vanilla RNN that'd be awesome.

Thanks and keep up the great work! really impressed at how much has been implemented already in this library, and how good the docs are already. I just need to start simple when it comes to RNNs and work my way up. If anyone has time I think a few more docs along these lines would make the library (whose strength in particular seems to be RNNs) even more accessible :)

**Comment by dwf** [comment]
CC @vdumoulin

**Comment by vdumoulin** [comment]
Point taken. I'm not very familiar with the content of `blocks.bricks.sequence_generators`, but to me it looks like you don't even need to worry about these bricks if you're interested in training a vanilla RNN (@rizar could you chime in on that?).

Would seeing how to implement something like regression on a time series (_i.e._ predict `y_1, ..., y_T` given `x_1, ..., x_T`) be informative enough for you? I can try to find a toy example and write a short tutorial section on how to train an RNN on that toy example.

If anyone has an idea for the toy example, I'm all ears!

**Comment by bartvm** [comment]
No need for sequence generators if you're just looking at RNNs. As the name suggests, you only need sequence generators if you're doing sequence-to-sequence stuff.

As far as RNN examples go, maybe some of the experiments from the original LSTM paper? Those are pretty simple e.g. learning to remember the first element of a sequence (so input is `x, noise, noise, ...` and target is `x`).

**Comment by mjwillson** [comment]
Hey, thanks for the response.

I'm not doing sequence-to-sequence stuff but I am training an RNN to predict the next token at every timestep as in an RNN language model. So I need a cost function which sums over the costs at each timestep when predicting the next word.

It looked like I might need something like SequenceGenerator to do that, together with SoftmaxEmmiter and LookupFeedback? I guess the problem for me is that SequenceGenerator seems  designed to do something a lot more general and it's not immediately clear how to use it for something simpler.

I'll have a look at the LSTM examples for now though and see if they help...

**Comment by mjwillson** [comment]
Another thing which it'd be good to have some docs on are the conventions for how multiple training examples for RNNs are (from what I can see) batched up into a rank 3 tensor. Is this mandatory i.e. to you have to build a network that operates simultaneously on an entire minibatch in a higher-order tensor? The docs don't seem to explicitly explain this. Assuming that is the case, the other thing that's not clear is how that works with variable sequence lengths. Are you expected to pad all sequences in a minibatch to the same maximum length and have your network ignore the blanks? any advice on this? I notice some code around "masks" in the SequenceEmmiter code.

**Comment by rizar** [comment]
Hey @mjwillson,

thank you for pushing us to work more on the documentation. There is an unfortunately abandoned PR #563 which I should finish.

If you want to train a language model, then sequence generators is what you need. It takes care of feeding outputs back to the network which you otherwise would have to implement yourself. In fact the markov chain example is exactly what you need.

Multiple training examples should be batched up into matrices if they consist of discrete data and 3d-tensors otherwise. Padding can be done by `Padding` from https://github.com/bartvm/fuel/blob/master/fuel/transformers/__init__.py. 

Let me know if I missed any of your questions. So far I guess you should follow the markov chain example:  it's a bit outdated but functional (at least the test passes)

**Comment by bartvm** [comment]
I'm quite sure you can train a language model without the need for sequence generators. Just take `x_1, ..., x_t` as the inputs and `x_2, ..., x_{t + 1}` as the targets. For each target word, the network has already seen all previous words as inputs, so it's properly conditioned.

**Comment by rizar** [comment]
@bartvm in the same fashion one can simply Theano and not use Blocks. For language modelling one has too add a softmax layer and an embedding matrix to the RNN and when one uses sequence generator he can get both by simply specifying feedback and emitter bricks. Also sequence generator can be used to build the generation computation graph which allows to qualitatively explore the language model.

**Comment by bartvm** [comment]
I don't see any analog with using Theano. As a first experiment it's completely reasonable to use the vanilla recurrent bricks first to get familiar with them before moving to the much more complicated sequence generator part of the library, especially considering the lack of documentation. Many people have difficulty using them.

**Comment by rizar** [comment]
Sure, I agree that implementing things yourself for the first time is very helpful. Similarly for full understanding of what's going on one might want to implement an RNN from scratch not using bricks. But apart from that language modelling with sequence generator is just less code, since this component was originally designed with this task in mind.

**Comment by mjwillson** [comment]
@rizar thanks for the pointers! I'll have a look at those docs-in-progress too.

On a second look at the markov example I think I can figure out how to use it. I was partly confused because I was expecting to see `.apply` called on the SequenceGenerator brick at some point as in the other examples in the docs. But it looks like you call .cost_matrix on it instead to apply it to theano variables for its input sequences, and then you need to compute a theano mean or other aggregate on that cost matrix to get something you can optimise.

Thanks for the pointer to Padder too. From looking at how these masks are used, I guess I'm just slightly concerned that this could lead to a bunch of redundant computation in cases where you have a few long sequences but most are short and are padded a lot. Is that right? Or perhaps theano is able to do some clever optimisations here to avoid computing hidden states for the masked padding entries?

**Comment by bartvm** [comment]
You're absolutely right about the redundant computation due to padding. There's no optimization applied in these cases. For large datasets where this is an issue, we do the following:

* Read N examples form the dataset
* Sort N examples by length
* Split the N examples up into M batches
* Pad the batches

If you choose a large enough N/small enough M, this will result in batches that are pretty similar in length, reducing the amount of redundant computation.  I started writing documentation on this for Fuel at one point, but haven't got to finishing it yet: https://github.com/bartvm/fuel/compare/translation_doc

**Comment by mjwillson** [comment]
OK, yeah sorting by length sounds like a decent strategy in that case. Could even call it "curriculum learning" and claim it as a benefit :)
